### PR TITLE
Add cleanup routine to free memory when args parsing failed

### LINKED ISSUE #267: cndpfwd: cleanup if we fail when parsing args
Difficulty: Easy to Medium

Some additional cleanup is needed to ensure that malloc'ed/mmap'd memory is freed when we return in error during argument parsing in: https://github.com/CloudNativeDataPlane/cndp/blob/main/examples/cndpfwd/parse-args.c

### PR DESCRIPTION
**Pull request by haoruan**
#### What type of PR is this?
/kind bugfix

#### What this PR does / why we need it:
In cndpfwd example, when args parsing failed, the memory is not freed correctly.
This PR add a cleanup routine to free allocated memory when error returns from argument parsing.

Sounds like other examples also have this issue, but this pr only targets cndpfwd example.

#### Which issue(s) this PR fixes:
Fixes #267 

Signed-off-by: Hao Ruan <hao.ruan@intel.com>
Signed-off-by: Xiaotian Chen <xiaotian.chen@intel.com>

### CODE REVIEW

**Review by KeithWiles — CHANGES_REQUESTED** [review]
Looking good, but have a few comments.

File: examples/cndpfwd/parse-args.c [hunk 1]
```diff
@@ -390,11 +443,21 @@ parse_args(int argc, char **argv, struct fwd_info *fwd)
...
-    if (!fwd->opts.no_metrics && (enable_metrics(fwd) || enable_uds_info(fwd)))
-        CNE_ERR_RET("*** Failed to start metrics support ***\n");
+    if (!fwd->opts.no_metrics) {
+        int ret = enable_metrics(fwd);
+        if (ret == 0) {
+            ret = enable_uds_info(fwd);
+            if (ret)
+                metrics_destroy();
+        }
+        if (ret)
```
[the remark this answers is not in the document]
Comment (haoruan) [reply examples/cndpfwd/parse-args.c:467]: In this case, we will miss the failure of `enable_ud_info`

File: examples/cndpfwd/parse-args.c [hunk 2]
```diff
@@ -294,6 +293,60 @@ print_usage(char *prog_name)
...
+
+        switch (fwd->pkt_api) {
+        case XSKDEV_PKT_API:
+            if (pd->xsk) {
+                xskdev_socket_destroy(pd->xsk);
+                pd->xsk = NULL;
+            }
+            break;
+        case PKTDEV_PKT_API:
+            // TODO: The init value of pd->lport is zero, it seems to be safe to call pktdev_close.
```
[the remark this answers is not in the document]
Comment (haoruan) [reply examples/cndpfwd/parse-args.c]: Updated

File: examples/cndpfwd/parse-args.c [hunk 3]
```diff
@@ -301,6 +304,58 @@ print_usage(char *prog_name)
...
+                pktmbuf_destroy(lport->umem->rinfo[i].pool);
+                lport->umem->rinfo[i].pool = NULL;
+            }
+            mmap_free(lport->umem->mm);
+            lport->umem->mm = NULL; /* Make sure we do not free this again */
+        }
+
+        struct fwd_port *pd = lport->priv_;
+
+        if (!pd) {
```
Comment (KeithWiles) [inline examples/cndpfwd/parse-args.c:326]: Remove single line {} for if, for, while, do, ...

Comment (haoruan) [reply examples/cndpfwd/parse-args.c]: Updated

**Pull request merged**